File: insertion_sort.c
// C program for insertion sort
#include <math.h>
#include <stdio.h>

// Function to sort an array using insertion sort
void insertion_sort(int arr[], int n)
{
	int i, key, j;
	for (i = 1; i < n; i++) /*loops from 2nd array element(arr[1]) 
								to last array element(arr[n-1]) to compare arr[i-1] & arr[i].
								If arr[i-1] > arr[i], then swap them*/
    {
		key = arr[i];
		j = i - 1;

		/* Move elements of arr[0..i-1], that are
		greater than key, to one position ahead
		of their current position */
		while (j >= 0 && arr[j] > key) {
			arr[j + 1] = arr[j];
			j = j - 1;
		}
		arr[j + 1] = key;
	}
}

// A function to print the sorted array of size n
void print_array(int arr[], int n)
{
	int i;
    printf("Array after sorting\n");
	for (i = 0; i < n; i++)
		printf("%d ", arr[i]);
	printf("\n");
}

//main function
int main()
{
	int n,i;
    printf("Enter the size of the array\n");
    scanf("%d",&n);//To take input the length of array from the user
	int arr[n];
    printf("Enter the array elements\n");
    //for loop to take input into the array
    for(i=0;i<n;i++)
    {
        scanf("%d",&arr[i]);
    }

	insertion_sort(arr, n);//calling the insertion_sort() function with arguments arr and n(here the array is passed as address)
	print_array(arr, n);//calling the print_array() function to print the array after sorting

	return 0;
}

File: insertionSort.java
// Insertion sort is a simple sorting algorithm that works similar to the way you sort playing cards in your hands.
// The array is virtually split into a sorted and an unsorted part. 
// Values from the unsorted part are picked and placed at the correct position in the sorted part.
// Algorithm 
// To sort an array of size n in ascending order: 
// 1: Iterate from arr[1] to arr[n] over the array. 
// 2: Compare the current element (key) to its predecessor. 
// 3: If the key element is smaller than its predecessor, compare it to the elements before. Move the greater elements one position up to make space for the swapped element.
// Java program for implementation of Insertion Sort
class InsertionSort {
	/*Function to sort array using insertion sort*/
	void sort(int arr[])
	{
		int n = arr.length;
		for (int i = 1; i < n; ++i) {
			int key = arr[i];
			int j = i - 1;

			/* Move elements of arr[0..i-1], that are
			greater than key, to one position ahead
			of their current position */
			while (j >= 0 && arr[j] > key) {
				arr[j + 1] = arr[j];
				j = j - 1;
			}
			arr[j + 1] = key;
		}
	}

	/* A utility function to print array of size n*/
	static void printArray(int arr[])
	{
		int n = arr.length;
		for (int i = 0; i < n; ++i)
			System.out.print(arr[i] + " ");

		System.out.println();
	}

	// Driver method
	public static void main(String args[])
	{
		int arr[] = { 12, 11, 13, 5, 6 };

		InsertionSort ob = new InsertionSort();
		ob.sort(arr);

		printArray(arr);
	}
} /* This code is contributed by Arpit Tripathi. */
